kmod/src/forest.c

```diff
@@ -801,23 +810,162 @@ int scoutfs_forest_oino_alloc(struct super_block *sb, u64 trans_seq)
 	finf->oino_bl = NULL;
+
+	schedule_delayed_work(&finf->oino_work, msecs_to_jiffies(OINO_WORK_DELAY));
+
+	return 0;
+}
+
+void scoutfs_oino_delete_work(struct work_struct *work)
+{
+	struct forest_info *finf = container_of(work, struct forest_info,
+						oino_work.work);
+	struct super_block *sb = finf->sb;
+	struct scoutfs_net_roots roots;
+	SCOUTFS_BTREE_ITEM_REF(iref);
+	struct scoutfs_log_trees lt = {{{ 0 }}};
+	struct scoutfs_bloom_block *bb;
+	struct scoutfs_bloom_block *first_bb = NULL;
+	struct scoutfs_block *bl;
+	struct scoutfs_block *first_bl = NULL;
+	struct scoutfs_key ltk;
+	int ret = 0;
+	int i;
+
+	if (finf->shutdown)
+		return;
+
+	scoutfs_inc_counter(sb, forest_oino_delete_work);
+
+	ret = scoutfs_client_get_roots(sb, &roots);
+	if (ret)
+		goto out;
+
+	scoutfs_key_init_log_trees(&ltk, 0, 0);
+	for (;; scoutfs_key_inc(&ltk)) {
+		ret = scoutfs_btree_next(sb, &roots.logs_root, &ltk, &iref);
+		if (ret == 0) {
+			if (iref.val_len == sizeof(lt)) {
+				ltk = *iref.key;
+				memcpy(&lt, iref.val, sizeof(lt));
+			} else {
+				ret = -EIO;
+			}
+			scoutfs_btree_put_iref(&iref);
+		}
+		if (ret < 0) {
+			if (ret == -ENOENT)
+				break;
+			goto out; /* including stale */
+		}
+
+		if (lt.oino_bloom_ref.blkno == 0)
+			continue;
+
+		bl = read_bloom_ref(sb, &lt.oino_bloom_ref);
+		if (IS_ERR(bl)) {
+			ret = PTR_ERR(bl);
+			goto out;
+		}
+
+		/* If more than one, merge later bloom filters into the first */
+		if (!first_bl) {
+			first_bl = bl;
+			first_bb = first_bl->data;
+		} else {
+			bb = bl->data;
+
+			/* Use oldest block's seq */
+			first_bb->hdr.seq = cpu_to_le64(
+				min(le64_to_cpu(bb->hdr.seq),
+				    le64_to_cpu(first_bb->hdr.seq)));
+
+			for (i = 0; i < SCOUTFS_FOREST_BLOOM_ARRAY_LEN; i++) {
+				first_bb->bits[i] |= bb->bits[i];
+			}
+
+			scoutfs_block_put(sb, bl);
+		}
+	}
+
+	if (first_bb) {
+		ret = scoutfs_scan_orphans(sb, first_bb);
+		if (ret > 0) {
+			schedule_work(&finf->oino_work.work);
+			ret = 0;
+		}
+
+		scoutfs_block_put(sb, first_bl);
+	}
+
+out:
+	if (ret == -ESTALE) {
+		scoutfs_inc_counter(sb, forest_oino_bloom_stale);
+		schedule_delayed_work(&finf->oino_work, msecs_to_jiffies(OINO_WORK_DELAY));
+	} else {
+
+		WARN_ON_ONCE(ret);
+	}
+}
+
+/*
+ * Use merged bloom filter to detect if an inode is likely open on another node.
+ */
+bool scoutfs_forest_oino_deletable(struct super_block *sb,
+				   __le64 le_ino,
+				   u64 orphan_seq,
+				   struct scoutfs_bloom_block *bb)
+{
+	int bloom_hits = 0;
+	u64 hash;
+	int i;
+
+	/* ino not deletable until orphan seq older than merged bloom's seq */
+	if (orphan_seq > le64_to_cpu(bb->hdr.seq)) {
+		scoutfs_inc_counter(sb, forest_oino_deletable_false_1);
+		return false;
+	}
+
+	/* ino not deletable until bloom says it's closed everywhere */
+	hash = scoutfs_hash64(&le_ino, sizeof(le_ino));
+	for (i = 0; i < SCOUTFS_FOREST_BLOOM_NRS; i++) {
+		u32 idx = (u32)hash % SCOUTFS_FOREST_BLOOM_BITS;
+		bloom_hits += !!test_bit_le(idx, bb->bits);
```

Comment:
Just break if test_bit_le() is false, then use i == _NRS to detect if they all passed.  The typical case is that unlinked files won't be open so we'll save some time.